--- origsrc/mono-2.6.4/mono/utils/mono-mmap.c	2010-04-22 10:37:45.000000000 -0500
+++ src/mono-2.6.4/mono/utils/mono-mmap.c	2010-06-28 15:08:50.048454600 -0500
@@ -376,12 +376,11 @@ mono_mprotect (void *addr, size_t length
 	if (flags & MONO_MMAP_DISCARD) {
 		/* on non-linux the pages are not guaranteed to be zeroed (*bsd, osx at least) */
 #ifdef __linux__
-		if (madvise (addr, length, MADV_DONTNEED))
+		if (posix_madvise (addr, length, POSIX_MADV_DONTNEED))
 			memset (addr, 0, length);
 #else
 		memset (addr, 0, length);
-		madvise (addr, length, MADV_DONTNEED);
-		madvise (addr, length, MADV_FREE);
+		posix_madvise (addr, length, POSIX_MADV_DONTNEED);
 #endif
 	}
 	return mprotect (addr, length, prot);
