--- origsrc/mono-2.6.4/mono/utils/mono-hash.c	2010-04-22 10:37:45.000000000 -0500
+++ src/mono-2.6.4/mono/utils/mono-hash.c	2010-06-28 15:53:56.949280500 -0500
@@ -98,16 +98,16 @@ static void mono_g_hash_mark (void *addr
 
 G_LOCK_DEFINE_STATIC (g_hash_global);
 
-#if defined(HAVE_NULL_GC)
-static GMemChunk *node_mem_chunk = NULL;
-#endif
 #if defined(HAVE_SGEN_GC)
 static MonoGHashNode *node_free_lists [4] = {NULL};
 static void *hash_descr = NULL;
 static GMemChunk *node_mem_chunk = NULL;
-#else
+#elif defined(HAVE_BOEHM_GC)
 static void *node_gc_descs [4] = {NULL};
 static MonoGHashNode *node_free_lists [4] = {NULL};
+#else
+static GMemChunk *node_mem_chunk = NULL;
+static MonoGHashNode *node_free_list = NULL;
 #endif
 
 #ifdef HAVE_SGEN_GC
